### PR TITLE
vfs: add format in vfs API

### LINKED ISSUE #8314: vfs: how to properly format?
#### Description
Current `vfs` does not contain a `format` function. `spiffs` as well as `littlefs` (in progress) integrations format the filesystem if an error occurs during `mount`. I think this behavior, even if it's handy to test and quickly implement user code, is not good for long-term. On the other hand `fatfs` integration seems to work differently, and in this case, if I am correct, the user must call directly the format function from fatFS.

I am wondering if we should add a `format` function in the `vfs_file_system_ops_t` structure to let the user format its filesystem.
Any other idea?

### PR DESCRIPTION
**Pull request by vincent-d**
### Contribution description

There was no generic way to format a file system so far. This PR add a `vfs_format` function in vfs API and implements it for spiffs.

### Issues/PRs references

Fixes #8314

### CODE REVIEW

**Comment by vincent-d** [comment]
Rebased and added littlefs format

**Comment by vincent-d** [comment]
Fixed unittests

**Comment by vincent-d** [comment]
Rebased.

Could someone take a look to this one?

**Comment by gebart** [comment]
Working on a review, will finish it tomorrow

**Comment by vincent-d** [comment]
@gebart thanks for the review. I'll address your comments.

To answer the question, I wondered if it should be part of vfs, and I came up with this because of its simplicity.

The use case we try to cover is indeed auto formatting when mounting fails. We have implemented a basic partition table and we can check and format at boot time if needed. At least the first boot, we need to format memory.

> If ROM usage becomes a problem we can add a compile time option to conditionally remove functions from the vfs_xxx_ops structs to be able to strip those functions from the binary.

+1

File: sys/vfs/vfs.c [hunk 1]
```diff
@@ -450,10 +490,16 @@ int vfs_mount(vfs_mount_t *mountp)
 int vfs_umount(vfs_mount_t *mountp)
 {
     DEBUG("vfs_umount: %p\n", (void *)mountp);
-    if ((mountp == NULL) || (mountp->mount_point == NULL)) {
+    int ret = check_mount(mountp);
+    if (ret == 0) {
```
[the remark this answers is not in the document]
Comment (vincent-d) [reply sys/vfs/vfs.c]: changed to a switch block, but I'm not sure to understand what you meant with the default.

File: sys/vfs/vfs.c [hunk 2]
```diff
@@ -426,14 +435,45 @@ int vfs_mount(vfs_mount_t *mountp)
...
+    }
+    mutex_unlock(&_mount_mutex);
+
+    if (mountp->fs->fs_op != NULL) {
+        if (mountp->fs->fs_op->format != NULL) {
+            return mountp->fs->fs_op->format(mountp);
+        }
+    }
+
+    return 0;
```
[the remark this answers is not in the document]
Comment (vincent-d) [reply sys/vfs/vfs.c]: changed to -ENOTSUP

**Comment by vincent-d** [comment]
Comments addressed

File: sys/vfs/vfs.c [hunk 3]
```diff
@@ -450,10 +491,16 @@ int vfs_mount(vfs_mount_t *mountp)
...
-    if ((mountp == NULL) || (mountp->mount_point == NULL)) {
+    int ret = check_mount(mountp);
+    switch (ret) {
+    case 0:
+        DEBUG("vfs_umount: not mounted\n");
+        mutex_unlock(&_mount_mutex);
         return -EINVAL;
+    case -EINVAL:
+        DEBUG("vfs_umount: invalid fs\n");
+        return ret;
```
Comment (gebart) [inline sys/vfs/vfs.c]: What I meant with default case was to add:
```c
default:
return -EINVAL;
```
to have as fallback if check_mount misbehaves.

File: sys/vfs/vfs.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (vincent-d) [reply sys/vfs/vfs.c]: OK, I guess it's better with the last version then.

File: sys/vfs/vfs.c (hunk 3 above)
Comment (vincent-d) [reply sys/vfs/vfs.c]: See last commit

**Comment by gebart** [comment]
please squash

**Comment by vincent-d** [comment]
> I am getting errors from littlefs during unit tests.

These errors happen when trying to mount an invalid file system, so they are "normal". We should silent these prints I guess. There are coming from littlefs directly.

**Comment by vincent-d** [comment]
Squashed

**Comment by gebart** [comment]
ACK, feel free to merge when Murdock agrees

**Comment by vincent-d** [comment]
> ACK, feel free to merge when Murdock agrees

Then go!

Thanks @gebart

**Pull request merged**